### PR TITLE
`Plagiarism checks`: Use file extensions for plagiarism view

### PR DESCRIPTION
**Pull request by magaupp**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I _adapted_ integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.

### Motivation and Context

The detection of binary files involves Java's `Files.probeContentType()` where the results are sometimes guessed from the filename. The results can not reliably detect source code files.

### Description

This PR the list of file extensions used by the code editor for the detection of binary files. This also makes one REST call unnecessary.

### Steps for Testing

Prerequisites:
- 1 Programming Exercise with supported plagiarism checks
- 2 identical submissions with text and binary files named accordingly

1. Run the plagiarism checks on the exercise
2. Select the resulting plagiarism case
3. Select a text file
4. Verify that the file's contents are shown
5. Select a binary file
6. Verify that `Binary file not rendered.` is shown

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

## Release Notes

- **New Features**
	- Introduced a new constant for text file extensions to enhance file type handling.

- **Bug Fixes**
	- Simplified logic for determining file types in the Text Submission Viewer component.

- **Refactor**
	- Renamed constants related to file extensions for clarity and consistency across components.
	- Removed unused methods for fetching file types and headers, streamlining the codebase.

- **Tests**
	- Updated test cases for the Text Submission Viewer component to reflect changes in file handling logic.

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/text-submission-viewer/text-submission-viewer.component.ts [hunk 1]
```diff
@@ -174,26 +175,24 @@ export class TextSubmissionViewerComponent implements OnChanges {
...
-                        this.loading = false;
-                        this.fileContent = this.insertMatchTokens(fileContent);
-                    },
-                    error: () => {
-                        this.loading = false;
-                    },
-                });
-            }
-        });
+        const fileExtension = file.split('.').last()!;
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/text-submission-viewer/text-submission-viewer.component.ts:178]: `split()` cannot possibly return an empty array here.

File: src/main/webapp/app/exercises/programming/shared/code-editor/file-browser/code-editor-file-browser.component.ts [hunk 2]
```diff
@@ -582,7 +582,7 @@ export class CodeEditorFileBrowserComponent implements OnInit, OnChanges, AfterV
      */
     private static shouldDisplayFileBasedOnExtension(fileName: string): boolean {
         const fileSplit = fileName.split('.');
-        return fileSplit.length === 1 || supportedTextFileExtensions.includes(fileSplit.pop()!);
+        return fileSplit.length === 1 || TEXT_FILE_EXTENSIONS.includes(fileSplit.pop()!);
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/webapp/app/exercises/programming/shared/code-editor/file-browser/code-editor-file-browser.component.ts:585]: split() cannot possibly return an empty array here.

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts [hunk 3]
```diff
@@ -39,9 +39,9 @@ export class AttachmentUnitFormComponent implements OnInit, OnChanges {
     isEditMode = false;
 
     // A human-readable list of allowed file extensions
-    readonly allowedFileExtensions = FILE_EXTENSIONS.join(', ');
+    readonly allowedFileExtensions = UPLOAD_FILE_EXTENSIONS.join(', ');
     // The list of file extensions for the "accept" attribute of the file input field
-    readonly acceptedFileExtensionsFileBrowser = FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
+    readonly acceptedFileExtensionsFileBrowser = UPLOAD_FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts:44]: For prepending a single character concatenation is more readable.

File: src/main/webapp/app/lecture/lecture-update.component.ts [hunk 4]
```diff
@@ -46,9 +46,9 @@ export class LectureUpdateComponent implements OnInit {
     faHandShakeAngle = faHandshakeAngle;
 
     // A human-readable list of allowed file extensions
-    readonly allowedFileExtensions = FILE_EXTENSIONS.join(', ');
+    readonly allowedFileExtensions = UPLOAD_FILE_EXTENSIONS.join(', ');
     // The list of file extensions for the "accept" attribute of the file input field
-    readonly acceptedFileExtensionsFileBrowser = FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
+    readonly acceptedFileExtensionsFileBrowser = UPLOAD_FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/webapp/app/lecture/lecture-update.component.ts:51]: For prepending a single character concatenation is more readable.

File: src/main/webapp/app/lecture/lecture-attachments.component.ts [hunk 5]
```diff
@@ -34,9 +34,9 @@ export class LectureAttachmentsComponent implements OnInit, OnDestroy {
     viewButtonAvailable: Record<number, boolean> = {};
 
     // A human-readable list of allowed file extensions
-    readonly allowedFileExtensions = FILE_EXTENSIONS.join(', ');
+    readonly allowedFileExtensions = UPLOAD_FILE_EXTENSIONS.join(', ');
     // The list of file extensions for the "accept" attribute of the file input field
-    readonly acceptedFileExtensionsFileBrowser = FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
+    readonly acceptedFileExtensionsFileBrowser = UPLOAD_FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/webapp/app/lecture/lecture-attachments.component.ts:39]: For prepending a single character concatenation is more readable.

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by raffifasaro — DISMISSED** [review]
Code lgtm

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts [hunk 6]
```diff
@@ -44,3 +44,301 @@ export const FILE_EXTENSIONS = [
...
     'odf',
 ];
+
+/**
+ * The list of file extensions that are readable in a file editor.
+ * Extensions are case-sensitive.
+ * Extensions are the part of the filename after the last dot.
+ * Files without dots are represented with their full filename.
+ */
+export const TEXT_FILE_EXTENSIONS = [
```
Comment (ole-ve) [inline src/main/webapp/app/shared/constants/file-extensions.constants.ts:54]: Also `.gradle` for `build.gradle` and `settings.gradle`?

**Review by ole-ve — DISMISSED** [review]
Reviewed code, added one smaller suggestion (which I'm not sure about)

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts (hunk 6 above)
Comment (magaupp) [reply src/main/webapp/app/shared/constants/file-extensions.constants.ts:54]: Yes, that probably belongs here. But I didn't intend to add any changes to this list in this PR.

**Review by SimonEntholzer — APPROVED** [review]
Code lgtm

**Review by raffifasaro — APPROVED** [review]
re-approval

**Review by ole-ve — APPROVED** [review]
Code after last approval

**Pull request merged**